Automattic/jetpack
# Site Settings API: Add `subscription_options` to site settings endpoint

#### Changes proposed in this Pull Request:

The proposed changes expose `subscription_options` site option to the site settings API at `site/%s/settings` (`GET`) and also allows this setting to be updated (`POST`).

Closes https://github.com/Automattic/wp-calypso/issues/71414

Having `subscription_options` site option available is required to be able to interact with it via Calypso UI, more specifically through the new Reading settings page we are building. This PR will unblock the following two tasks:

* https://github.com/Automattic/wp-calypso/issues/70414
* https://github.com/Automattic/wp-calypso/issues/70415

#### Other information:


#### Jetpack product discussion


* project thread: pdDOJh-17o-p2
* p1671626436207709-slack-C02TCEHP3HA


#### Does this pull request change what data or activity we track or use?


No, it does not.

#### Testing instructions:

1. Use the [script below](https://github.com/Automattic/jetpack/pull/28036#issuecomment-1361259623) to sync the changes to your sandbox.
2. Sandbox `public-api.wordpress.com`.
3. When testing with Atomic and Jetpack-connected sites, the build changes need to be synced to those sites as well. More info: PCYsg-eg0-p2 and pb5gDS-1rQ-p2
4. Use the Developer Console to test the `GET` request (`/sites/$site/settings`):

![Markup on 2023-01-03 at 10:26:36](https://user-images.githubusercontent.com/25105483/210330495-188cb34b-296d-4f7e-ae80-984edb94c927.png)

5. Use cURL to test `POST` requests:

As a first step, you can copy the raw cURL request by doing a POST request using the Developer Console:

![Markup on 2023-01-03 at 10:37:30](https://user-images.githubusercontent.com/25105483/210333846-e706c9e1-98c1-4717-88bc-40add242a86a.png)

and then paste it into terminal with desired option values to be set, e.g.:

![Markup on 2023-01-03 at 10:52:00](https://user-images.githubusercontent.com/25105483/210334198-6ab7d419-abf6-484e-87f3-dd3657735436.png)

Once the request is sent, the API should respond as follows:

![Markup on 2023-01-03 at 10:53:11](https://user-images.githubusercontent.com/25105483/210334353-8abc35c6-8cd7-4acb-9ba3-32413bc53f24.png)

and the new option values should be saved.

The `POST` should let through values that only include plain text, line breaks and anchor tags. The accepted keys are `invitation` and `comment_follow`. Example of a correct option in JSON format:

```json
"subscription_options": {
	"invitation": "text 1",
	"comment_follow": "text 2"
},
```

If needed, values can be also changed through textfields in the WP Admin UI at `wp-admin/options-reading.php`.

ℹ️ Please note that if the `subscription_options` hasn't been touched through admin yet (`wp-admin/options-reading.php`) e.g. when the site is new, the option will be empty. On the WP Admin side, the default values are extracted from `Subscription_Mailer::get_default_settings()` - which we don't have access to here. Related discussion: p1671718726381529/1671717628.207919-slack-C02TCEHP3HA. → This issue will be resolved separately: https://github.com/Automattic/wp-calypso/issues/71711.
